crash-handler-process/platforms/upload-window-win.cpp

```diff
@@ -0,0 +1,413 @@
+
+	This program is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 2 of the License, or
+	(at your option) any later version.
+
+	This program is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+******************************************************************************/
+
+#include <windows.h>
+#include <CommCtrl.h>
+#include <thread>
+
+#include "../logger.hpp"
+#include "upload-window-win.hpp"
+
+UploadWindow* UploadWindow::instance = NULL;
+
+#define CUSTOM_CLOSE_MSG (WM_USER + 1)
+#define CUSTOM_PROGRESS_MSG (WM_USER + 2)
+
+#define CUSTOM_CATCHED_CRASH (WM_USER + 3)
+#define CUSTOM_SAVE_STARTED (WM_USER + 4)
+#define CUSTOM_SAVED_DUMP (WM_USER + 5)
+#define CUSTOM_SAVING_DUMP_FAILED (WM_USER + 6)
+#define CUSTOM_UPLOAD_STARTED (WM_USER + 7)
+#define CUSTOM_UPLOAD_FINISHED (WM_USER + 8)
+#define CUSTOM_UPLOAD_FAILED (WM_USER + 9)
+#define CUSTOM_UPLOAD_CANCELED (WM_USER + 10)
+
+#define CUSTOM_REQUESTED_CLICK (WM_USER + 12)
+
+LRESULT CALLBACK FrameWndProc(HWND hwnd, UINT msg, WPARAM wParam, LPARAM lParam)
+{
+	switch (msg) {
+	case WM_CLOSE:
+		return 0;
+	case WM_NCCREATE:
+	case WM_CREATE: {
+		break;
+	}
+	default: {
+		return UploadWindow::getInstance()->WndProc(hwnd, msg, wParam, lParam);
+	}
+	}
+	return DefWindowProc(hwnd, msg, wParam, lParam);
+}
+
+LRESULT UploadWindow::WndProc(HWND hwnd, UINT msg, WPARAM wParam, LPARAM lParam)
+{
+	switch (msg) {
+	case WM_DESTROY:
+		PostQuitMessage(0);
+		break;
+	case CUSTOM_CLOSE_MSG:
+		DestroyWindow(hwnd);
+		break;
+	case CUSTOM_CATCHED_CRASH: {
+		SetWindowText(upload_label_hwnd, L"Crash catched. Start saving a memory dump to a file?");
+		EnableWindow(ok_button_hwnd, true);
+		EnableWindow(cancel_button_hwnd, true);
+		SendMessage(progresss_bar_hwnd, PBM_SETBARCOLOR, 0, RGB(49, 195, 162));
+		SendMessage(progresss_bar_hwnd, PBM_SETBKCOLOR, 0, RGB(49, 49, 49));
+		SendMessage(progresss_bar_hwnd, PBM_SETRANGE32, 0, 100);
+		SendMessage(progresss_bar_hwnd, PBM_SETPOS, 0, 0);
+		ShowWindow(progresss_bar_hwnd, SW_SHOW);
+		break;
+	}
+	case CUSTOM_PROGRESS_MSG: {
+		double progress = ((double)bytes_sent) / ((double)total_bytes_to_send / 100.0);
+		PostMessage(progresss_bar_hwnd, PBM_SETPOS, (int)progress, 0);
+		swprintf(upload_progress_message, L"%lld / %lld Kb", bytes_sent / 1024, total_bytes_to_send / 1024);
+		SetWindowText(upload_label_hwnd, upload_progress_message);
+		break;
+	}
+	case CUSTOM_SAVED_DUMP: {
+		SetWindowText(upload_label_hwnd, L"Memory dump saved successfully. Continue with upload?");

```

Comment:
I'm wondering why we should ask the user again if they want to upload. In my opinion once they said yes on sharing additional data to developers, we should save and upload.